aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws.s3;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.relocated.com.google.common.hash.HashCode;
+import org.apache.iceberg.relocated.com.google.common.hash.HashFunction;
+import org.apache.iceberg.relocated.com.google.common.hash.Hashing;
+import org.apache.iceberg.util.LocationUtil;
+
+/**
+ * This location provider provides data locations that are optimized for S3 performance. Both
+ * General Purpose buckets and Directory buckets will see better throughput and autoscaling behavior
+ * than using the generic ObjectStoreLocationProvider.
+ *
+ * <p>The data location is resolved as follows. Data files are written directly at this path with no
+ * other intermediate directories created.
+ *
+ * <ol>
+ *   <li>{@link TableProperties#WRITE_DATA_LOCATION}
+ *   <li><code>tableLocation + "/data"</code>
+ * </ol>
+ *
+ * The data file is placed immediately under the data location. Partition names are <b>not</b>
+ * included. The data filename is prefixed with a 24-character binary hash, which ensures that files
+ * written to S3 are equally distributed across many prefixes in the S3 bucket.
+ *
+ * <p>For example, with tableLocation <code>s3://my-bucket/my-table</code>, an example data file
+ * could look like <code>
+ * s3://my-bucket/my-table/data/011101101010001111101000-00000-0-5affc076-96a4-48f2-9cd2-d5efbc9f0c94-00001.parquet
+ * </code>.
+ */
+public class S3LocationProvider implements LocationProvider {
```

Comment:
Why not just include this in the ObjectStoreLocationProvider as an option to choose between base2 and base32?  It seems like we could include most of this in just the `computeHash` function.  That might also allow for choosing whether to include partition context.

Reply:
That was our original plan but with the partition values and other trailing directories removed, it turns into a very S3 specific location provider due to reasons mentioned in the above comment. I agree that if it was just the base2 optimization then it would fit into the ObjectStoreLocationProvider with one extra table property probably.